Std/Data/Array/Match.lean

```diff
@@ -0,0 +1,126 @@
+/-
+Copyright (c) 2023 F. G. Dorais. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: F. G. Dorais
+-/
+import Std.Data.Array.Basic
+import Std.Data.Array.Lemmas
+
+namespace Array
+
+/-- Prefix table for the Knuth-Morris-Pratt matching algorithm
+
+  This is an array of the form `t = [(x₀,n₀), (x₁,n₁), (x₂, n₂), ...]` where for each `i`, `nᵢ` is
+  the length of the longest proper prefix of `xs = [x₀,x₁,...,xᵢ]` which is also a suffix of `xs`.
+-/
+structure PrefixTable (α : Type _) extends Array (α × Nat) where
+  /-- Validity condition to help with termination proofs -/
+  valid : (h : i < toArray.size) → toArray[i].2 ≤ i
+
+instance : Inhabited (PrefixTable α) where
+  default := ⟨#[], fun .⟩
+
+/-- Returns the size of the prefix table -/
+abbrev PrefixTable.size (t : PrefixTable α) := t.toArray.size
+
+/-- Returns the pattern array of the prefix table -/
+abbrev PrefixTable.pattern (t : PrefixTable α) : Array α := t.toArray.map Prod.fst
+
+/-- Transition function for the KMP matcher
+
+  Assuming we have an input `xs` with a suffix that matches the pattern prefix `t.pattern[:len]`
+  where `len : Fin (t.size+1)`. Then `xs.push x` has a suffix that matches the pattern prefix
+  `t.pattern[:t.step x len]`. If `len` is as large as possible then `t.step x len` will also be
+  as large as possible.
+-/
+def PrefixTable.step [BEq α] (t : PrefixTable α) (x : α) : Fin (t.size+1) → Fin (t.size+1)
```

Comment:
Could `[BEq \a]` be a variable?

Reply:
I'm not sure what you mean. Just making it an implicit parameter?

Reply:
Sorry: write `variable [BEq \a]` at the top of the file, so it isn't in every type signature.

Reply:
@semorrison I try to avoid `variable` wherever possible in std. In this case I think it's fine as is.

Reply:
I just reverted the change but I must admit having no personal preference either way.